### DISCUSSION TITLE
All About Transformer Inference | How To Scale Your Model

### DISCUSSION THREAD

**Discussion by jacobaustin123** [jax-ml/scaling-book#9] [General]
Serving big models!

**Comment by clu0** [comment]
In the first figure for sharding KV comparing MHA and MQA, I think the locations of the reduce-scatter and all-gather are reversed.

**Comment by rohanmahajan1993** [comment]
Had question and was trying to understand more. 

1. "For prefill,   since we’re doing self-attention, so this simplifies to 
 This is great because it means the arithmetic intensity of attention during prefill is 
Θ(T). That means it’s quite easy to be compute-bound for attention. As long as our batch size and sequence length are both fairly large, we’ll be fine!"

why do we need our batch size also to be fairly large, it seems like the equation is defined by token length; so isn't that sufficient

2. 
"Why is this, conceptually? Mainly, we’re compute-bound in linear portions of the model because the parameters (the memory bandwidth-heavy components) are reused for many batch items. However, every batch item has its own KV cache, so a bigger batch size means more KV caches. We will almost always be memory bound here unless the architecture is adjusted aggressively."

another question why the batch size effects this being memory bound; doesn't batch size cancel out here in the equations you're showing;  I get the point that t=1 for generation, and since we have to fetch the kv cache stuff for s, we're still doing kind of half the memory work for that but we lose the s*t multiplication for the flopers inference.

**Reply by jacobaustin123** [reply]
1. I think this should basically just say "as long as our sequence length is fairly large". I think before this was support to say "if our batch size is large our MLPs are good, and if our sequence length is long, our attention is good".

2. Could you restate this slightly? The idea is, for our MLPs, we load one matrix for B entries, so the FLOPs / byte loaded increases with batch size (since our activations are small). For attention, as we increase B, we load B times more KV caches, so our intensity doesn't improve.

**Reply by rohanmahajan1993** [reply]
On 2, it was  totally my bad. I think initially I misread and thought the text was implying that increasing batch size decreases the intensity of the attention but in reality the point it was  that it doesn't change the attention part.

**Comment by mmcclean-aws** [comment]
I have a question on the Generation section and the calculations around time for HBM loading. It seems to only consider the weight loading time and not also the KV cache loading time as well. Does this need to be added to the calculation ?

**Reply by jacobaustin123** [reply]
Do you mean under "Distributing Inference Over Multiple Accelerators"? I think in general here we're looking at the matmul vs. ICI comms overlap, which is unrelated to the KV caches. Those will always be memory-bound and more model parallelism will always improve things

**Reply by mmcclean-aws** [reply]
Am referring to this formula. The time for HBM comms is only considering model weights and not KV cache loading time. For long sequences the KV cache can be larger than the model weights so should also be taken into consideration
![Screenshot 2025-03-06 at 1 52 33 PM](https://github.com/user-attachments/assets/06242aeb-71b3-4d6a-b2a2-57eb1f477b14)

**Reply by jacobaustin123** [reply]
Think about what we're trying to calculate here. We want to know how much model parallelism we can do before ici communication during our matmuls takes longer than weight loading. At that point, we hit a limit and more model parallelism won't reduce the time our matmuls take. It _will_ reduce the time KV loading takes, so if all you care about is latency, you can do even more.

So I guess the point is, this is the rule for how much model parallelism you can do before you start hitting diminishing returns on throughput at a fixed batch size. But you can go beyond it if you're willing to take a throughpout hit.

**Reply by YuzeLou** [reply]
Hi Jacob, when we hit the limit of model parallelism, why would it further decrease latency? Isn't latency here max{T_hbm_comm, T_ici_comm}?

Should we also add M_mesh (number of mesh axis) to the denominator of T_ici_comm?

**Comment by Lhongpei** [comment]
About the part of "exact number will differ on the kind of quantization and hardware".If we use int8 to store parameters but bf16 for computation, why is the result of critical 120？ According to the following formulas:

$$
\frac{2BDF}{2BD + 2DF + 2BF} \geq \frac{\text{TPU FLOPs/s}}{\text{HBM Bandwidth}} = \frac{1.97E+14}{8.20E+11} = 240
$$

For int8, the communication will decrease to half, then the critical B will be 480?

**Reply by jacobaustin123** [reply]
Not quite. If we simplify, we end up with 2B > 240, or B > 120.

**Comment by aasthavar** [comment]
For the formula:
```
FFW params = d_model^2 * ffw_multiplier * 3 * n_layers
```
I am unsure where the `*3*` came from? Shouldn't it be just `*2*`?

For one layer, given:
- **Input Projection:**  
  - `W_in.shape = [d_model, d_mlp]`  
  - `b_in.shape = [d_mlp]`
- **GeLU**
   - no weights
- **Output Projection:**  
  - `W_out.shape = [d_mlp, d_model]`  
  - `b_out.shape = [d_model]`  

Total params for one layer:
```
params = (d_model * d_mlp) + (d_mlp * d_model) 
       = (d_model^2) * ffw_multiplier * 2
```

Total params across all layers:
```
total_params = params * n_layers 
             = (d_model^2) * ffw_multiplier * 2 * n_layers
```

**Reply by jacobaustin123** [reply]
So this is a little confusing in this book because we assume what you describe in Section 5 for simplicity. 

However, GeGeLU, the gated variant described in https://arxiv.org/abs/2002.05202v1 has two up-projection matrices whose output is pointwise multiplied. There’s a note on this in Section 4.

**Comment by aasthavar** [comment]
Shouldn't there be positional embedding matrix (of shape = [max_seq_len, d_model]) as well, along with ffw+vocab+attn params?

**Reply by jacobaustin123** [reply]
Sometimes. Many models today use some kind of parameter free positional embedding, or where the number of params is very small, eg ROPE, T5-style relative positional embedding, etc.

Either way, the number is so small it essentially doesn’t matter.

**Comment by VijayGKR** [comment]
For the newly defined model we're using in the worked problems, should the number of heads be 16 instead of 32, or qk dim be 128 instead of 256? Right now, d_qkv * n_heads is 8192, instead of matching d_model at 4096.

**Reply by jacobaustin123** [reply]
Probably yes, maybe worth updating. Still, it's not a strict rule that d_qkv * n_heads = d_model. The main thing that matters is the number n_kv heads

**Comment by andrewkchan** [comment]
> **Note on ICI bounds for generation:** during training we want to be compute-bound, so our rooflines look at when our ICI comms take longer than our FLOPs. However, during generation, if we’re memory bandwidth bound by parameter loading, we can increase model sharding beyond this point and improve latency at a minimal throughput cost.

I don't understand why there would be any throughput cost when increasing model sharding to improve latency in generation while memory-bound. 
1. If we're holding batch size constant, then doesn't improving latency strictly improve throughput? 
2. Otherwise, IIUC increasing the number of accelerators we shard over yields more HBM, which means we can increase batch size. As long as we are memory-bound, we can actually therefore improve throughput while holding latency constant, no?

**Reply by jacobaustin123** [reply]
I agree that increasing the number of chips _in theory_ should leave throughput constant if we don't change the batch size. If we're memory-bound on 8 chips, and then we increase to 16 and we're still memory-bound, we're using 2x more chips for 2x more throughput (since the weight loading time is halved), so totally constant. 

In practice we tend to suffer some downside from trying to do more model sharding, like some ICI overhead that isn't totally free, since it's rare for us to be fully memory-bound. So I think you have the right idea, it's just that in practice it's likely to have some overhead.

**Reply by andrewkchan** [reply]
I see. In that case I'm confused about what definition of decode throughput we're using. I was assuming `throughput = batch_size / latency`, where a decrease in latency with batch size constant must improve throughput. Maybe worth clarifying?

**Reply by jacobaustin123** [reply]
I see, I typically mean batch size / latency / num chips, i.e. tokens/sec/chip. I'll clarify, but I think that's the quantity that matters.

**Comment by zhiyisun** [comment]
Minor typo in the first "Takeaway"

Takeaway: Transformer matmuls are compute-bound ***iff*** the per-replica token batch size is greater than

**Reply by jacobaustin123** [reply]
This was intended, cause iff is sometimes used in math to mean "if and only if". It's a kind of niche usage though.

**Comment by mayun8172** [comment]
In the multi query attention full algorithm,

The step 7 is confused to me 

```
6. O[B_Z, S, K_Y, M] = Q[B_Z, K_Y, M, H] *_H K[B_Z, S, K_Y, H]
7. O[B_Z, S, K, M] = Softmax_S(O[B_Z, S, K_Y])
```

Do we should softmax along Y instead ?

```
7. O[B_Z, S, K, M] = Softmax_Y(O[B_Z, S, K_Y, M])
```

**Comment by tylsun** [comment]
For problem 1. Shouldn't the KV cache size be 2*L∗K∗H per token? 1 for K, 1 for V.

**Reply by jacobaustin123** [reply]
Yes totally missed on this one. Will fix.

**Reply by SimonBerens** [reply]
Doesn't seem like this was fixed?

**Reply by jacobaustin123** [reply]
Fixed now, thanks for the reminder.

**Comment by IrishWhiskey** [comment]
The computed FLOPs for the attention layer are `4BTSD`. How come it differs from what's shown under "Global FLOPs and Params Calculation" in chaper 4? I understand that this chapter assumes there's only one head (no H), but the math still differs quite a bit.

**Reply by jacobaustin123** [reply]
In part 4 from only the dot product attention we have effectively 12BT^2NH during training. We’re assuming NH=D, and B=1. In this section we have 4STD for inference. We expect training to have 3x the inference FLOPs, and letting S=T we have the same thing. Hope that helps.

**Reply by IrishWhiskey** [reply]
In part 4, the FLOPs for the dot product attention is `12BTD(N + K)H`. I still don't get how this becomes `12BT^2NH`. Why does `K` disappear? Where does the `T^2` come from?

**Reply by jacobaustin123** [reply]
So T^2 = ST where we assume the query and key context is the same (self-attention). But more importantly, here we’re just talking about the dot product part of the attention, excluding the QKV projections which are just matmuls. That’s the key thing you’re missing, that those get lumped in with the linear part.

**Reply by IrishWhiskey** [reply]
Ah got it! It makes sense now. Thank you :)

**Comment by rschreyer** [comment]
The symbols alpha_hbm = C / W_hbm were brought in without introduction.   What are those intended to represent?

**Reply by jacobaustin123** [reply]
I think these are both defined in section 5: https://jax-ml.github.io/scaling-book/training.

**Reply by rschreyer** [reply]
Right, there they are:
C = per-chip flops
W_hbm still doesn't seem to be explicitly defined, but W_ici is inter-chip bandwidth, so W_hbm is clearly HBM bandwidth

**Comment by ammar1510** [comment]
If B<sub>crit</sub> is so high for hardware (240), why isn't compute cost traded for faster memory bandwidth in chip design? to bring it closer to a ratio of 1? that way during inference wouldn't have to rely on a large batch size to be compute bound.

**Reply by jacobaustin123** [reply]
I'll start by saying that I'm not a hardware expert, so others may have better answers to this question. One answer is that that critical batch size _is_ achieved in the vast majority of use cases (training and prefill) and LLMs are relatively new and thus not explicitly optimized for. There are efforts to design LLM-optimized chips that may make this tradeoff (although it would likely mean a separate chip for generation vs. prefill). Also, memory bandwidth is harder to add than compute.

**Comment by vedantjh2** [comment]
in Q5.4, the num params is stated as 32.2e9. I believe this is a typo if the num active params previously calculated was 31.2e9

**Reply by jacobaustin123** [reply]
Good catch. This is now fixed!

**Comment by wangdelp** [comment]
Thanks for the great work! @jacobaustin123 On question 2:
- Isn't the KVCache size has a multiplier of 2, b/c we have both K and V, therefore the per token size is  262kB as we computed in question 1?
- I wonder why the num_heads*qkv_dim=32*256=8192 is twice of the d_model 4096 in the table? Shouldn't this number always match?

**Reply by jacobaustin123** [reply]
1. Thank you for catching this, a total typo on my part. Fixed now.
2. This rule of thumb is often but not always true. In fact often it can be smaller than d\_model, and sometimes it can be more, although less often. Don't read too much into it here.

**Comment by nimarez** [comment]
nit: In the "What about attention?" section, the hover-over tooltip footnote has a typo. Whese -> These

**Reply by jacobaustin123** [reply]
Fixed! Will update github in a few hours

**Comment by nimarez** [comment]
Thank you for the great blog. I might be misunderstanding this, but in the pop quiz under the “Theoretical estimates for LLM latency and throughput”, I believe the 16x multiplier for bandwidth for 4 tokens only applies if we’re both sharding the weights and the KV cache, not if they’re replicated right?

**Reply by jacobaustin123** [reply]
Yes, in general it's pretty important to fully shard both, although it can be hard because of the number of KV heads.

**Comment by he-ye** [comment]
In the answer of Question 5(4), do you mean "2*(number of activated parameters)*T" instead of "2ND"? It is a bit confusing to use N and D here, as in the context above, D usually means d_model and N usually means #heads.

**Reply by jacobaustin123** [reply]
Yes, 6ND is just sort of the canonical way of writing even though it's confusing. I'll change it.

**Comment by lhy-loveworld** [comment]
Regarding `Bcrit =C/Whbm ⋅(bits per param/bits per activation)=β⋅αhbm`, this doesn't sound right when the param is in fp16 and activation is fp8, because the matmul will still be done in fp16, which means `Bcrit =C/Whbm`, right?

**Comment by lhy-loveworld** [comment]
3 potential typos(a bit nit picky):
1. In question 7: `and find when it will outperform traditional 3D model sharding?` shouldn't it be `traditional 1D model sharding`?
2. In appendix B: `Modifying the 2D sharding for attention (instead of sharding the hidden dimension, we shard the n_heads dimension)`. aren't we already sharding the n_heads dimension in 1D sharding for attention? What's the other dimension we are sharding for attention?
3. In appendix D speculative decoding: `For normal autoregressive sampling the token/s is the same as the step time.`, token/s should be the reciprocal of the step time :)

**Comment by kabir-vats** [comment]
I really like this book.

There's a calculation error in the last part of question 7's solution. It's written that $128 \times \left(\frac{3}{4}\right)^2 = 81$, but the answer is $72$.

**Comment by flufcloud** [comment]
In the section 'What about attention?' it says that attention requires us to 
> Read the 𝐾𝑉 cache, which is a pair of bf16[B, S, D] tensors from HBM.

But earlier it says that during prefill we *write* the KV cache. I think the math works out the same either way, but just to check my understanding, it should be write/read for prefill/generation right? 

Thanks for the great resource!

**Reply by jacobaustin123** [reply]
Good catch, typically prefill is done in chunks, so you read a previous chunk you've already prefilled, do the attention, then write the new result back.

**Reply by flufcloud** [reply]
Thank you!

**Comment by msisovic** [comment]
In the solution to question 2 it is assumed that we can perfectly shard the attention matrices across the 16 TPUs as well. The book hasn't mentioned how this is done so far, as attention was largely overlooked in the previous chapter, I can for now only infer one way how this can be done from the sharded dot product attention algorithm provided in this chapter, something like:

- $W_Q[D, N_{YZ}, H]$
- $W_K[D, K_Y, H]$
- $W_V[D, K_Y, H]$
- $W_O[N_{YZ}, H, D]$

The issue is, here K is less than the number of TPUs at hand, what do we usually do in this case, do we just avoid sharding further or shard along some other dimension, TP-style?

**Reply by msisovic** [reply]
For ppl reading this in the future: I found an answer to my question in a later chapter (chapter 9): "We’re Megatron sharding along the head dimension.", so exactly what I asked about is indeed possible.

**Comment by gordicaleksa** [comment]
Amazing book, thanks for making it! :)

I believe the formulas for Question 7 [2D model sharding] are incorrect, after doing all-gather across y dimension you'd end up with non-contiguous chunks of `In` matrix doing matmul against 2D chunks of `Win` matrix which would give incorrect partial results.

The all-gather should go over x instead, i assume?

I unfortunately failed to attach my visualizations to this comment but hopefully the above makes it clear.

**Reply by msisovic** [reply]
Hey, I think the algorithm is actually correct, but I may have missed your point. If you gather over X, I think it would in that case you would actually end up with non-contiguous chunks of In.

Let me explain. With a `[XYZ]` layout the single index is numbered row-major over the three mesh axes, `index = (X·M_Y + Y)·M_Z + Z` (where M_Y, M_Z are the sizes of the Y and Z mesh axes), so X is the slowest-varying (outer) coordinate, Z the fastest (inner), and contiguity is inherited from the innermost axis outward. Gathering away a *suffix* of the sharding (the trailing axes YZ in this case), keeps reuniting adjacent runs into a contiguous block sharded over the remaining outer prefix X, matching the weights' contraction sharding, whereas gathering away X first would break that suffix rule and leave each chip holding one index per X-slice, a set strided by M_Y·M_Z.

I made a toy example in a colab notebook where a one-dimensional array is sharded over a 2d chip mesh, and allgathered over both axes: https://colab.research.google.com/drive/1B2FvrQBi0FtxdH8KnyWWb961mUk-PF4i?usp=sharing

**Comment by sritee** [comment]
Thanks for the great resource. 

Quick question on the $T_{ICI}$ comms calculation in 'Note on ICI bounds for generation' section: shouldn't it be $4BD / W_{ici}$ rather than $2BD / W_{ici}$, since the all-gather will be reduce-scatter + all-reduce (plus bf16 = 2 bytes), as discussed in the Training section?

**Reply by sritee** [reply]
typo: i mean all-reduce will be reduce-scatter + all-gather

**Reply by sritee** [reply]
it's possibly this might be because even in calculation of t_hbm, there's a shift to now use float8 rather than bf16 as before for generation.
